internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package http
+
+import (
+	"context"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/atomic"
+)
+
+type downloadProgressReporter struct {
+	sourceURI   string
+	interval    time.Duration
+	warnTimeout time.Duration
+	length      float64
+
+	downloaded atomic.Int
+	started    time.Time
+
+	progressObservers []progressObserver
+}
+
+func newDownloadProgressReporter(sourceURI string, timeout time.Duration, length int, progressObservers ...progressObserver) *downloadProgressReporter {
+	interval := time.Duration(float64(timeout) * downloadProgressIntervalPercentage)
+	if interval == 0 {
+		interval = downloadProgressMinInterval
+	}
+
+	return &downloadProgressReporter{
+		sourceURI:         sourceURI,
+		interval:          interval,
+		warnTimeout:       time.Duration(float64(timeout) * warningProgressIntervalPercentage),
+		length:            float64(length),
+		progressObservers: progressObservers,
+	}
+}
+
+func (dp *downloadProgressReporter) Write(b []byte) (int, error) {
+	n := len(b)
+	dp.downloaded.Add(n)
+	return n, nil
+}
+
+func (dp *downloadProgressReporter) Report(ctx context.Context) {
+	started := time.Now()
+	dp.started = started
+	sourceURI := dp.sourceURI
+	length := dp.length
+	interval := dp.interval
+
+	go func() {
```

Comment:
very unnecessary optimization from my side: no need to spin up goroutine or initiate tickers if progressObservers is []. as we're having at least log now, i consider this just a small optimization from component side of view.

Reply:
Good point, will add, thanks!

Reply:
Added in da06f8c9b.